### PR TITLE
Add support to convert to uniform axis in `interpolate_on_axis`

### PR DESCRIPTION
**Pull request by ericpre**
### Progress of the PR
- [x] Add support to convert to uniform axis in `interpolate_on_axis`,
- [x] Fixes `FunctionalDataAxis.convert_to_uniform_axis` (`AttributeError`),
- [n/a] update docstring (if appropriate),
- [n/a] update user guide (if appropriate),
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/upcoming_changes/README.rst)),
- [x] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [x] add tests,
- [x] ready for review.

### Minimal example of the bug fix or the new feature
```python
import hyperspy.api as hs
import numpy as np

dict0 = {'expression': 'x**2', 'size': 500}
s = hs.signals.Signal1D(np.arange(500), axes=[dict0])

# s = hs.signals.Signal1D(np.arange(100), axes=)
s2 = s.interpolate_on_axis("uniform", -1, inplace=False)

hs.plot.plot_spectra([s, s2], legend=["FunctionalAxis", "Interpolated"])
```

### CODE REVIEW

**Comment by Attolight-NTappy** [comment]
Hi @ericpre 

Definitely a very useful feature! I know you didn't ask for my feedback, but I would just like to point out a few thing if you don't mind 😁 .
- If I read code correctly, the way you implement linearization does not minimize error with respect to the original axis. For axes without inflection points (i.e. most non-uniform axes), it has a tendency to maximize error at the axis center. I issued a small code example and figure to illustrate this below
- I suspect that `convert_to_uniform_axis` will break on descending axes (due to the behavior of `self.low_value`).

Cheers, 

Nicolas

```python
import numpy as np
import matplotlib.pyplot as plt

#Non uniform axis
idx = np.arange(1024)
axis = 321.4 + idx*0.654 - 0.246e-3*idx**2
#Linear 1 (mean diff)
steps = axis[1:] - axis[:-1]
scale = np.mean(steps)
offset = axis[0]
linear_1 = offset+idx*scale
#Linear 2 (polyfit)
scale,offset = np.polyfit(idx,axis,1)
linear_2 = offset + idx*scale

plt.figure()
plt.plot(axis,label='non-uniform')
plt.plot(linear_1,'k--',label='meandiff')
plt.plot(linear_2,'k-.',label='polyfit')
plt.xlabel('index')
plt.ylabel('axis value')
plt.legend()
```
image

**Comment by ericpre** [comment]
Thank you @Attolight-NTappy, it makes sense to change to polyfit!

The idea of this PR is to make easier to use `interpolate_on_axis` by re-sampling on a uniform axis and without having to specify the axis. I wasn't thinking to touch on the `convert_to_uniform_axis` method, but got sidetracked when I noticed that there is still an error when converting a `DataAxis` with a uniform axis give to a `UniformDataAxis` and this could be avoided. This is why I changed slightly the way the scale and offset are estimated.

**Comment by ericpre** [comment]
Actually, when using in combination with `interpolate_on_axis`, it is useful to keep the bounds of the axis the same to avoid extrapolating outside the data outside the original range.

```python
import hyperspy.api as hs

s = hs.data.luminescence_signal(uniform=False)
s2 = s.interpolate_on_axis("uniform", -1, inplace=False)

hs.plot.plot_spectra([s, s2], legend=["FunctionalAxis", "Interpolated (identical bounds)"], drawstyle='steps-mid')


new_axis = s.axes_manager[-1].copy()
new_axis.convert_to_uniform_axis(keep_bounds=False, log_scale_error=False)

s3 = s.interpolate_on_axis(new_axis, -1, inplace=False)
hs.plot.plot_spectra([s, s3], legend=["FunctionalAxis", "Interpolated (polyfit)"], drawstyle='steps-mid')

s2_scale = s2.axes_manager[-1].scale
s3_scale = s3.axes_manager[-1].scale
print("scale with identical bounds", s2_scale)
print("scale using polyfit", s3_scale)

print("Relative difference: ", abs(s3_scale - s2_scale) / max(s3_scale, s2_scale))
```

image

image

```python
scale with identical bounds 0.0038831643648140246
scale using polyfit 0.0033323404163766376
Relative difference:  0.14184924888281608
```

File: hyperspy/axes.py [hunk 1]
```diff
@@ -729,18 +729,80 @@ def update_from(self, axis, attributes):
...
+        <Axes manager, axes: (|1024)>
+                    Name |   size |  index |  offset |   scale |  units
+        ================ | ====== | ====== | ======= | ======= | ======
+        ---------------- | ------ | ------ | ------- | ------- | ------
+                  Energy |   1024 |      0 |     1.1 |  0.0033 |     eV
+
+        """
+        indices = np.arange(self.size)
+        if keep_bounds:
+            scale = np.mean(np.diff(self.axis))
```
Comment (Attolight-NTappy) [inline hyperspy/axes.py]: Mathematically equivalent to 
`scale = (self.axis[-1] - self.axis[0])/len(self.axis)`

which is cheaper computationally, and (to me) shows more explicitly that bounds are preserved. What you think?

**Comment by Attolight-NTappy** [comment]
> Actually, when using in combination with interpolate_on_axis, it is useful to keep the bounds of the axis the same to avoid extrapolating outside the data outside the original range.

Ah that's a good point! I didn't think of it that way. Also, if the data is re-interpolated on new values, I guess it does not matter that the new axis points are far away from the original values. Good catch

File: hyperspy/axes.py (hunk 1 above)
Comment (ericpre) [reply hyperspy/axes.py]: Done!

**Comment by ericpre** [comment]
Yes, that's right.

File: hyperspy/axes.py (hunk 1 above)
Comment (Attolight-NTappy) [reply hyperspy/axes.py]: Nice! Also nice that you realized I had off-by-one error ^^

**Comment by ericpre** [comment]
@jlaehne, since this would be relevant to you, will you be interested in having a look at this PR?

**Comment by jlaehne** [comment]
Yes, I'll try to find a moment this week.

**Comment by ericpre** [comment]
In relation to support of non-uniform axes in ROI (as mentioned in https://github.com/hyperspy/hyperspy/pull/3418#discussion_r1704424981 for example): I think that if we have an easy way to convert to uniform axis (as in this PR), we should recommend this in favour of implementing support of non-uniform for all ROI, because I don't think that this is worth the effort. What do you think?

File: hyperspy/axes.py [hunk 2]
```diff
@@ -729,18 +729,80 @@ def update_from(self, axis, attributes):
...
-        """Convert to an uniform axis."""
-        scale = (self.high_value - self.low_value) / self.size
+    def convert_to_uniform_axis(self, keep_bounds=True, log_scale_error=True):
+        """
+        Convert to an uniform axis.
+
+        Parameters
+        ----------
+        keep_bounds : bool
+            If ``True``, the first and last value of the axis will not changed.
```
Comment (jlaehne) [inline hyperspy/axes.py]:
```suggestion
            If ``True``, the first and last value of the axis will not be changed.
```

File: hyperspy/axes.py [hunk 3]
```diff
@@ -729,18 +729,80 @@ def update_from(self, axis, attributes):
             any_changes = True
         return any_changes
 
-    def convert_to_uniform_axis(self):
-        """Convert to an uniform axis."""
-        scale = (self.high_value - self.low_value) / self.size
+    def convert_to_uniform_axis(self, keep_bounds=True, log_scale_error=True):
+        """
+        Convert to an uniform axis.
```
Comment (jlaehne) [inline hyperspy/axes.py:734]: Add a note that this function touches the axis only and not the data - with a hint that `interpolate_on_axis` can be used if the data should be adjusted accordingly.

File: hyperspy/axes.py [hunk 4]
```diff
@@ -729,18 +729,80 @@ def update_from(self, axis, attributes):
...
+
+        Parameters
+        ----------
+        keep_bounds : bool
+            If ``True``, the first and last value of the axis will not changed.
+            The new scale is estimated by averaging the :func:`numpy.diff` of
```
Comment (jlaehne) [inline hyperspy/axes.py]: It is in fact not using `numpy.diff` - though the effect of `scale = (self.axis[-1] - self.axis[0]) / (self.size - 1)` should indeed be the same as `scale = np.mean(np.diff(self.axis))`.

File: hyperspy/axes.py [hunk 5]
```diff
@@ -729,18 +729,80 @@ def update_from(self, axis, attributes):
...
+            the axis. If ``False``, the scale and offset are calculated using
+            :meth:`numpy.polynomial.polynomial.Polynomial.fit`, which minimises
+            the scale difference over the whole axis range but the bounds of
```
Comment (jlaehne) [inline hyperspy/axes.py:744]: For an axis with a continuously changing interval as in the example, the latter shifts the range quite significantly!

File: hyperspy/axes.py [hunk 6]
```diff
@@ -729,18 +729,80 @@ def update_from(self, axis, attributes):
...
         d = self.get_axis_dictionary()
         axes_manager = self.axes_manager
-        del d["axis"]
-        if len(self.axis) > 1:
-            scale_err = max(self.axis[1:] - self.axis[:-1]) - scale
-            _logger.warning("The maximum scale error is {}.".format(scale_err))
+        if "axis" in d:
+            del d["axis"]
+        if len(self.axis) > 1 and log_scale_error:
+            scale_err = max(self.axis - scale * indices + offset)
```
Comment (jlaehne) [inline hyperspy/axes.py]: Shouldn't it be `- offset`? Now it gives an error that is larger than `2 * offset`.

File: hyperspy/signal.py [hunk 7]
```diff
@@ -3400,17 +3400,21 @@ def crop(self, axis, start=None, end=None, convert_units=False):
...
         ----------
         new_axis : :class:`hyperspy.axes.UniformDataAxis`,
-        :class:`hyperspy.axes.DataAxis` or :class:`hyperspy.axes.FunctionalDataAxis`
+        :class:`hyperspy.axes.DataAxis`, :class:`hyperspy.axes.FunctionalDataAxis`
+        or str
             Axis which replaces the one specified by the ``axis`` argument.
             If this new axis exceeds the range of the old axis,
             a warning is raised that the data will be extrapolated.
+            If ``"uniform"``, convert the axis specified by the ``axis``
+            parameter to a uniform axis.
```
Comment (jlaehne) [inline hyperspy/signal.py]:
```suggestion
            parameter to a uniform axis with the same number of data points.
```

File: hyperspy/signal.py [hunk 8]
```diff
@@ -3427,28 +3431,63 @@ def interpolate_on_axis(self, new_axis, axis=0, inplace=False, degree=1):
...
+        >>> s = hs.data.luminescence_signal(uniform=False)
+        >>> new_axis = s.axes_manager[-1].copy()
+        >>> new_axis.convert_to_uniform_axis(keep_bounds=False)
+        >>> s3 = s.interpolate_on_axis(new_axis, -1, inplace=False)
+        >>> hs.plot.plot_spectra(
+        ...     [s, s3],
+        ...     legend=["FunctionalAxis", "Interpolated (polyfit)"],
+        ...     drawstyle='steps-mid',
+        ... )
+        <Axes: xlabel='Energy (eV)', ylabel='Intensity'>
```
Comment (jlaehne) [inline hyperspy/signal.py:3470]: Running this second example gives a really weird result, as the noise is extrapolated over a huge distance, as mentioned above, `keep_bounds=False` does not really make sense for such a signal.

I would just leave out the second example here.

File: hyperspy/tests/signals/test_interpolate_on_axis.py [hunk 9]
```diff
@@ -121,6 +121,24 @@ def test_interpolate_error_navigate(self):
...
 
 
+def test_interpolate_convert_to_uniform():
+    s = signals.Signal1D(np.arange(100))
+    axis = s.axes_manager[-1]
+    axis.convert_to_non_uniform_axis()
+    assert not s.axes_manager[-1].is_uniform
+    s2 = s.interpolate_on_axis("uniform", -1, inplace=False)
+    assert s2.axes_manager[-1].is_uniform
+    np.testing.assert_allclose(s.data, s2.data)
```
Comment (jlaehne) [inline hyperspy/tests/signals/test_interpolate_on_axis.py:131]: assert that `axis` and the new axis of `s2` are same?

**Review by jlaehne — APPROVED** [review]

**Comment by jlaehne** [comment]
> In relation to support of non-uniform axes in ROI (as mentioned in [#3418 (comment)](https://github.com/hyperspy/hyperspy/pull/3418#discussion_r1704424981) for example): I think that if we have an easy way to convert to uniform axis (as in this PR), we should recommend this in favour of implementing support of non-uniform for all ROI, because I don't think that this is worth the effort. What do you think?

Indeed, at least for the moment that is a sensible approach.

Also, I would adapt the error message when trying to rebin a non-uniform axis, noting that one should first interpolate to a uniform axis and can then use rebin.

File: hyperspy/axes.py (hunk 4 above)
Comment (jlaehne) [reply hyperspy/axes.py]: Only saw the related comment above now, but then the docstring should be adapted

File: hyperspy/signal.py (hunk 8 above)
Comment (ericpre) [reply hyperspy/signal.py:3470]: I have left it and removed the "keep_bounds=False".

**Comment by ericpre** [comment]
@jlaehne do you want to have another look?

File: hyperspy/axes.py [hunk 10]
```diff
@@ -729,18 +729,92 @@ def update_from(self, axis, attributes):
...
+        """
+        Convert to an uniform axis.
+
+        Parameters
+        ----------
+        keep_bounds : bool
+            If ``True``, the first and last value of the axis will not be changed.
+            The new scale is calculated by substracting the last value by the first
+            value and dividing by the number of intervals.
+            the axis. If ``False``, the scale and offset are calculated using
```
Comment (jlaehne) [inline hyperspy/axes.py]:
```suggestion
            If ``False``, the scale and offset are calculated using
```

File: hyperspy/axes.py [hunk 11]
```diff
@@ -729,18 +729,92 @@ def update_from(self, axis, attributes):
...
+            :meth:`numpy.polynomial.polynomial.Polynomial.fit`, which minimises
+            the scale difference over the whole axis range but the bounds of
+            the axis can change. Default is ``True``.
```
Comment (jlaehne) [inline hyperspy/axes.py]:
```suggestion
            the axis can change (in some cases quite significantly, in particular when the
            interval width is changing continuously). Default is ``True``.
```

File: hyperspy/axes.py [hunk 12]
```diff
@@ -729,18 +729,92 @@ def update_from(self, axis, attributes):
...
+                  Energy |   1024 |      0 |     1.1 |  0.0033 |     eV
+
+
+        See Also
+        --------
+        hyperspy.api.signals.BaseSignal.interpolate_on_axis
+
+        Notes
+        -----
+        The function only convert the axis type and doesn't interpolate
```
Comment (jlaehne) [inline hyperspy/axes.py]:
```suggestion
        The function only converts the axis type and doesn't interpolate
```

**Comment by ericpre** [comment]
The failure is related to latest numpy release (2.1.0)

**Pull request merged**